ambry-router/src/main/java/com.github.ambry.router/GetOperation.java

```diff
@@ -0,0 +1,214 @@
+   */
+  Callback<T> getCallback() {
+    return operationCallback;
+  }
+
+  /**
+   * The exception associated with this operation if it failed; null otherwise.
+   * @return exception associated with this operation if it failed; null otherwise.
+   */
+  Exception getOperationException() {
+    return operationException;
+  }
+
+  /**
+   * Return the result of the operation.
+   * @return the operation result.
+   */
+  T getOperationResult() {
+    return operationResult;
+  }
+
+  /**
+   * Return the blob id string
+   * @return return the blob id string
+   */
+  String getBlobIdStr() {
+    return blobId.getID();
+  }
+
+  /**
+   * returns whether the operation has completed.
+   * @return whether the operation has completed.
+   */
+  boolean isOperationComplete() {
+    return operationCompleted;
+  }
+
+  /**
+   * Return the {@link MessageFormatFlags} to associate with the requests for this operation.
+   * @return the {@link MessageFormatFlags} to associate with the requests for this operation.
+   */
+  abstract MessageFormatFlags getOperationFlag();
+
+  /**
+   * For this operation, create and populate get requests to send out.
+   * @param requestRegistrationCallback the {@link RequestRegistrationCallback} to call for every request that gets
+   *                                    created as part of this poll operation.
+   */
+  abstract void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback);
+
+  /**
+   * Handle the given {@link ResponseInfo} received for a request that was sent out.
+   * @param responseInfo the {@link ResponseInfo} to be handled.
+   */
+  abstract void handleResponse(ResponseInfo responseInfo);
+
+  /**
+   * Process the given {@link ServerErrorCode} and set operation status accordingly.
+   * @param errorCode the {@link ServerErrorCode} to process.
+   */
+  void processServerError(ServerErrorCode errorCode) {
+    switch (errorCode) {
+      case Blob_Deleted:
+        logger.trace("Requested blob was deleted");
+        setOperationException(new RouterException("Server returned: " + errorCode, RouterErrorCode.BlobDeleted));
+        break;
+      case Blob_Expired:
+        logger.trace("Requested blob is expired");
+        setOperationException(new RouterException("Server returned: " + errorCode, RouterErrorCode.BlobExpired));
+        break;
+      case Blob_Not_Found:
+        logger.trace("Requested blob was not found on this server");
+        setOperationException(new RouterException("Server returned: " + errorCode, RouterErrorCode.BlobDoesNotExist));
+        break;
+      default:
+        setOperationException(
+            new RouterException("Server returned: " + errorCode, RouterErrorCode.UnexpectedInternalError));
+        break;
+    }
+  }
+
+  /**
+   * Set the exception associated with this operation.
+   * A {@link ServerErrorCode#Blob_Deleted} or {@link ServerErrorCode#Blob_Expired} error overrides any other
```

Comment:
Good point, but I think that is already being handled to the extent it can be. The chances of a blob being expired in some replicas and deleted in certain others is very slim, but even if that happens, here we ensure that the first such error that is reported is returned.